### PR TITLE
[FEATURE] Reusing the Pulsar message deduplication

### LINKED ISSUE #957: [FEATURE] Support idempotence with max.in.flight.requests.per.connection > 1
In KAFKA-5494, Kafka supported idempotence with `MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION > 1`, the client will auto-config the `MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION=5`, In KoP we should manual setting the `MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION=1` when we use idempotent producer, otherwise, it will disorder.

See `IdempotentProducerTest`
```java
Properties producerProperties = newKafkaProducerProperties();
producerProperties.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, "true");
producerProperties.put(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION, 1);
```

In KoP, supporting `MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION > 1` when using idempotent producer Is hard, since KoP uses async message publishing, is hard to add lock in this method. We should find a way to support it.

### PR DESCRIPTION
**Pull request by Demogorgon314**
Fixes #957 #1007

## Motivation
Currently, the KoP use Kafka's implantation to check duplicate message, but it is hard to support `MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION > 1`.

However, we should reuse the Pulsar message duplication check in KoP by mapping `baseSequence` to `sequenceId` and `lastSequence` to `highestSequenceId`.

Pulsar is using producer name to persist sequenced, in KoP we want to follow the Kafka behavior. So we need to use a name role to build a producer name.

Because Kafka will reuse PID when the transaction ID is the same but will increase the producer Enoch. So we need to ensure the producer name is not the same.

So the producer name role is `PID_PREFIX-{producerId}-{producerEpoch}`.

## Modifications
Support `MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION > 1`, and reuse Pulsar message deuplication.

### CODE REVIEW

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/MessagePublishContext.java [hunk 1]
```diff
@@ -60,17 +81,29 @@ public void completed(Exception exception, long ledgerId, long entryId) {
             offsetFuture.completeExceptionally(exception);
         } else {
             if (log.isDebugEnabled()) {
-                log.debug("Success write topic: {}, ledgerId: {}, entryId: {}"
+                log.debug("Success write topic: {}, producerName {} ledgerId: {}, entryId: {}"
                         + " And triggered send callback.",
-                    topic.getName(), ledgerId, entryId);
+                    topic.getName(), this.producerName, ledgerId, entryId);
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/MessagePublishContext.java:86]: It's better to keep the code style. `topic` and `producerName` are both fields of `MessagePublishContext`, but only `producerName` is referenced with `this.` prefix.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/AppendRecordsContext.java [hunk 2]
```diff
@@ -45,11 +45,12 @@ private AppendRecordsContext(Recycler.Handle<AppendRecordsContext> recyclerHandl
...
-                                           final RequestStats requestStats,
-                                           final Consumer<Integer> startSendOperationForThrottling,
-                                           final Consumer<Integer> completeSendOperationForThrottling,
-                                           final Map<TopicPartition, PendingTopicFutures> pendingTopicFuturesMap) {
+    public static AppendRecordsContext get(
+            final KafkaTopicManager topicManager,
+            final RequestStats requestStats,
+            final Consumer<Integer> startSendOperationForThrottling,
+            final Consumer<Integer> completeSendOperationForThrottling,
+            final Map<TopicPartition, PendingTopicFutures> pendingTopicFuturesMap) {
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/AppendRecordsContext.java:53]: Revert unnecessary changes.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/storage/ProducerStateManagerTest.java [hunk 3]
```diff
@@ -14,20 +14,16 @@
 package io.streamnative.pulsar.handlers.kop.storage;
 
 import static org.testng.Assert.assertEquals;
-import static org.testng.Assert.assertFalse;
-import static org.testng.Assert.assertThrows;
 import static org.testng.Assert.assertTrue;
-import static org.testng.Assert.fail;
+import static org.testng.AssertJUnit.fail;
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/storage/ProducerStateManagerTest.java:18]: Why do you use `AssertJUnit`?

**Comment by BewareMyPower** [comment]
Add the `doc-required` label since we need to enable deduplication at Pulsar side.

File: tests/src/test/java/io/streamnative/pulsar/handlers/kop/TransactionTest.java [hunk 4]
```diff
@@ -76,12 +77,12 @@ protected void cleanup() throws Exception {
         };
     }
 
-    @Test(timeOut = 1000 * 10, dataProvider = "produceConfigProvider")
+    @Test(timeOut = 1000 * 100, dataProvider = "produceConfigProvider")
```
Comment (BewareMyPower) [inline tests/src/test/java/io/streamnative/pulsar/handlers/kop/TransactionTest.java:80]: Why do you increase the timeout?

**Review by BewareMyPower — COMMENTED** [review]
Overall LGTM. I've left some comments, PTAL. I'll review again after comments are addressed.

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/ReplicaManager.java [hunk 5]
```diff
@@ -111,8 +111,12 @@ public PartitionLog getPartitionLog(TopicPartition topicPartition, String namesp
                         .thenAccept(offset -> addPartitionResponse.accept(topicPartition,
                                 new ProduceResponse.PartitionResponse(Errors.NONE, offset, -1L, -1L)))
                         .exceptionally(ex -> {
+                            Errors errors = Errors.forException(ex.getCause());
+                            if (errors == Errors.UNKNOWN_SERVER_ERROR) {
+                                errors = Errors.KAFKA_STORAGE_ERROR;
+                            }
```
Comment (BewareMyPower) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/ReplicaManager.java:117]: Could you explain why do you treat the `UNKNOWN_SERVER_ERROR` as `KAFKA_STORAGE_ERROR` here?

File: kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java [hunk 6]
```diff
@@ -275,53 +262,136 @@ private void publishMessages(final Optional<PersistentTopic> persistentTopicOpt,
...
                     producerStateManager.updateTxnIndex(completedTxn, lastStableOffset);
                     producerStateManager.completeTxn(completedTxn);
                 });
+
                 appendFuture.complete(offset);
             } else {
                 log.error("publishMessages for topic partition: {} failed when write.", fullPartitionName, e);
                 requestStats.getMessagePublishStats().registerFailedEvent(
                         time.nanoseconds() - beforePublish, TimeUnit.NANOSECONDS);
-                appendFuture.completeExceptionally(Errors.KAFKA_STORAGE_ERROR.exception());
```
Comment (Demogorgon314) [inline kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java:311]: @BewareMyPower See here, all exceptions will be treated as `Errors.KAFKA_STORAGE_ERROR`, I don't want to change previous behavior.

**Review by BewareMyPower — APPROVED** [review]
